lib/activerecord/debug_errors/ext/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -4,32 +4,54 @@
 module ActiveRecord
   module DebugErrors
     module DisplayMySQLInformation
-      # Rails 7 or later never calls ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter#execute
-      # cf. https://github.com/rails/rails/pull/43097
-      if ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter.private_method_defined?(:raw_execute)
-        method_name = :raw_execute
-      else
-        method_name = :execute
+      # For Rails 6.0, 6.1, 7.0.
+      if ActiveRecord.version < Gem::Version.new("7.1.0")
+        # Rails 7 or later never calls ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter#execute
+        # cf. https://github.com/rails/rails/pull/43097
+        if ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter.private_method_defined?(:raw_execute)
+          method_name = :raw_execute
+        else
+          method_name = :execute
+        end
+        define_method(method_name) do |*args, **kwargs|
+          super(*args, **kwargs)
+        rescue ActiveRecord::Deadlocked
+          handle_deadlocked
+          raise
+        rescue ActiveRecord::LockWaitTimeout
+          handle_lock_wait_timeout
+          raise
+        end
+        private method_name if method_name == :raw_execute
+      elsif ActiveRecord.version >= Gem::Version.new("7.1.0")
```

Comment:
As commented in https://github.com/abicky/activerecord-debug_errors/pull/3/commits/5f9a8c927f3051444a2e41b0d67a5b95bffda0fc, there is a case when neither `raw_execute` nor `execute` is called in Rails 7.1.

I'm not sure if this way is the best, I tried to use `translate_exception_class` as a hook for Rails 7.1.

Reply:
I think the code you presented is quite neat!

I think, I'm probably misunderstanding something big (and my implementation has based on the misunderstanding), but could you please explain to me why your code still works in Rails 7.1 for further studies? (and why the original code doesn't work.)

Reply:
As you can see below, `ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter#raw_execute` is not implemented, whereas the subclasses have the method `raw_execute`, so appending the module to the subclasses works expectedly:

```
% bundle exec irb
irb(main):001> require "active_record"
=> true
irb(main):002> require "active_record/connection_adapters/abstract_mysql_adapter"
=> true
irb(main):003> require "active_record/connection_adapters/mysql2_adapter"
=> true
irb(main):004> require "active_record/connection_adapters/trilogy_adapter"
=> true
irb(main):005> show_source ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter#raw_execute

From: /Users/arabiki/ghq/src/github.com/abicky/activerecord-debug_errors/gemfiles/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.3.3/lib/active_record/connection_adapters/abstract/database_statements.rb:530

        def raw_execute(sql, name, async: false, allow_retry: false, materialize_transactions: true)
          raise NotImplementedError
        end

=> nil
irb(main):006> show_source ActiveRecord::ConnectionAdapters::Mysql2Adapter#raw_execute

From: /Users/arabiki/ghq/src/github.com/abicky/activerecord-debug_errors/gemfiles/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.3.3/lib/active_record/connection_adapters/mysql2/database_statements.rb:96

          def raw_execute(sql, name, async: false, allow_retry: false, materialize_transactions: true)
            log(sql, name, async: async) do
              with_raw_connection(allow_retry: allow_retry, materialize_transactions: materialize_transactions) do |conn|
                sync_timezone_changes(conn)
                result = conn.query(sql)
                verified!
                handle_warnings(sql)
                result
              end
            end
          end

=> nil
irb(main):007> show_source ActiveRecord::ConnectionAdapters::TrilogyAdapter#raw_execute

From: /Users/arabiki/ghq/src/github.com/abicky/activerecord-debug_errors/gemfiles/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.3.3/lib/active_record/connection_adapters/trilogy/database_statements.rb:44

          def raw_execute(sql, name, async: false, allow_retry: false, materialize_transactions: true)
            log(sql, name, async: async) do
              with_raw_connection(allow_retry: allow_retry, materialize_transactions: materialize_transactions) do |conn|
                sync_timezone_changes(conn)
                result = conn.query(sql)
                verified!
                handle_warnings(sql)
                result
              end
            end
          end

=> nil
```

Reply:
Ah, I got it now. Thanks for your explanation!